modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/SleepTimer.kt

```diff
@@ -8,41 +8,78 @@ import android.os.Build
+    private val analyticsTracker: AnalyticsTrackerWrapper,
+    @ApplicationContext private val context: Context,
+) {
     companion object {
-        private var sleepTimeMs: Long? = null
+        private const val MIN_TIME_TO_RESTART_SLEEP_TIMER_IN_MINUTES = 5
+        private const val TIME_KEY = "time"
+        private const val END_OF_EPISODE_VALUE = "end_of_episode"
     }
 
-    fun sleepAfter(mins: Int, onSuccess: () -> Unit) {
+    private var sleepTimeMs: Long? = null
+    private var lastSleepAfterTime: Duration? = null
+    private var lastTimeSleepTimeHasFinished: Duration? = null
+    private var lastEpisodeUuidAutomaticEnded: String? = null
+
+    fun sleepAfter(minutes: Int, onSuccess: () -> Unit) {
         val time = Calendar.getInstance().apply {
             timeInMillis = System.currentTimeMillis()
-            add(Calendar.MINUTE, mins)
+            add(Calendar.MINUTE, minutes)
         }
         if (createAlarm(time.timeInMillis)) {
+            lastSleepAfterTime = minutes.toDuration(DurationUnit.MINUTES)
+            cancelAutomaticSleepOnEpisodeEndRestart()
             onSuccess()
         }
     }
 
-    fun addExtraTime(mins: Int) {
+    fun addExtraTime(minutes: Int) {
         val currentTimeMs = sleepTimeMs
         if (currentTimeMs == null || currentTimeMs < 0) {
             return
         }
         val time = Calendar.getInstance().apply {
             timeInMillis = currentTimeMs
-            add(Calendar.MINUTE, mins)
+            add(Calendar.MINUTE, minutes)
         }
         createAlarm(time.timeInMillis)
     }
+    fun restartSleepTimerIfApplies(currentEpisodeUuid: String, isSleepTimerRunning: Boolean, onRestartSleepAfterTime: () -> Unit, onRestartSleepOnEpisodeEnd: () -> Unit) {
+        lastTimeSleepTimeHasFinished?.let { lastTimeHasFinished ->
+            val diffTimeInMillis = System.currentTimeMillis() - lastTimeHasFinished.inWholeMilliseconds
+            val diffInMinutes = diffTimeInMillis / (1000 * 60) // Convert to minutes
+
+            if (diffInMinutes < MIN_TIME_TO_RESTART_SLEEP_TIMER_IN_MINUTES && !lastEpisodeUuidAutomaticEnded.isNullOrEmpty() && currentEpisodeUuid != lastEpisodeUuidAutomaticEnded) {
+                onRestartSleepOnEpisodeEnd()
+                analyticsTracker.track(PLAYER_SLEEP_TIMER_RESTARTED, mapOf(TIME_KEY to END_OF_EPISODE_VALUE))
+            } else if (diffInMinutes < MIN_TIME_TO_RESTART_SLEEP_TIMER_IN_MINUTES && lastSleepAfterTime != null && !isSleepTimerRunning) {
+                lastSleepAfterTime?.let {
+                    analyticsTracker.track(PLAYER_SLEEP_TIMER_RESTARTED, mapOf(TIME_KEY to it * 60)) // Convert to seconds
+                    sleepAfter(it.inWholeMinutes.toInt(), onRestartSleepAfterTime)
+                }
+            }
+        }
+    }
+    fun setEndOfEpisodeUuid(uuid: String) {
+        lastEpisodeUuidAutomaticEnded = uuid
+        lastTimeSleepTimeHasFinished = System.currentTimeMillis().toDuration(DurationUnit.MILLISECONDS)
```

Comment:
[Kotlin provides](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-duration/#companion-object-properties) useful extensions for these conversions.

```kotlin
1.days
10.seconds
378.milliseconds
```